bigquery/storage_iterator.go

```diff
@@ -0,0 +1,235 @@
+	ctx  context.Context
+
+	schema  Schema
+	decoder *arrowDecoder
+	records chan arrowRecordBatch
+
+	session *readSession
+}
+
+type arrowRecordBatch []byte
+
+func newStorageRowIteratorFromTable(ctx context.Context, table *Table) (*RowIterator, error) {
+	md, err := table.Metadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rs, err := table.c.rc.sessionForTable(ctx, table)
+	if err != nil {
+		return nil, err
+	}
+	it, err := newStorageRowIterator(ctx, rs, md.NumRows)
+	if err != nil {
+		return nil, err
+	}
+	it.arrowIterator.schema = md.Schema
+	return it, nil
+}
+
+func newStorageRowIteratorFromJob(ctx context.Context, job *Job, totalRows uint64) (*RowIterator, error) {
+	cfg, err := job.Config()
+	if err != nil {
+		return nil, err
+	}
+	qcfg := cfg.(*QueryConfig)
+	if qcfg.Dst == nil {
+		// TODO: script job ?
+		return nil, fmt.Errorf("nil job destination table")
+	}
+	return newStorageRowIteratorFromTable(ctx, qcfg.Dst)
+}
+
+func newRawStorageRowIterator(ctx context.Context, rs *readSession) (*arrowIterator, error) {
+	arrowIt := &arrowIterator{
+		ctx:     ctx,
+		session: rs,
+		records: make(chan arrowRecordBatch, 10000),
+		errs:    make(chan error, 1),
+	}
+	if rs.bqSession == nil {
+		err := rs.start()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return arrowIt, nil
+}
+
+func newStorageRowIterator(ctx context.Context, rs *readSession, totalRows uint64) (*RowIterator, error) {
+	arrowIt, err := newRawStorageRowIterator(ctx, rs)
+	if err != nil {
+		return nil, err
+	}
+	it := &RowIterator{
+		ctx:           ctx,
+		arrowIterator: arrowIt,
+		TotalRows:     totalRows,
+		rows:          [][]Value{},
+	}
+	it.nextFunc = nextFuncForStorageIterator(it)
+	it.pageInfo = &iterator.PageInfo{
+		Token:   "",
+		MaxSize: int(totalRows),
+	}
+	return it, nil
+}
+
+func nextFuncForStorageIterator(it *RowIterator) func() error {
+	return func() error {
+		if len(it.rows) > 0 {
+			return nil
+		}
+		arrowIt := it.arrowIterator
+		record, err := arrowIt.next()
+		if err == iterator.Done {
+			if len(it.rows) == 0 {
+				return iterator.Done
+			}
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		rows, err := arrowIt.decoder.decodeArrowRecords(record)
+		if err != nil {
+			return err
+		}
+		it.rows = rows
+		return nil
+	}
+}
+
+func (it *arrowIterator) init() error {
+	if it.decoder != nil { // Already initialized
+		return nil
+	}
+
+	bqSession := it.session.bqSession
+	if bqSession == nil {
+		return errors.New("read session not initialized")
+	}
+
+	streams := bqSession.Streams
+	if len(streams) == 0 {
+		return iterator.Done
+	}
+
+	if it.schema == nil {
+		meta, err := it.session.table.Metadata(it.ctx)
+		if err != nil {
+			return err
+		}
+		it.schema = meta.Schema
+	}
+
+	decoder, err := newArrowDecoderFromSession(it.session, it.schema)
+	if err != nil {
+		return err
+	}
+	it.decoder = decoder
+
+	go func() {
+		it.wg.Wait()
+		close(it.records)
+		it.done = true
+	}()
+
+	for _, readStream := range streams {
```

Comment:
You may want a semaphore here rather than spinning up N goroutines immediately for each reported stream, so you only have so many workers in flight at once.  Work through the list with a smaller number of active processStream instances